### PR TITLE
Add CTRL+ Mouse Scroll to scroll through last active stack

### PR DESCRIPTION
**Pull request by DragaDoncila**
# Description
Add event and binding for scrolling through stack using CTRL+Mouse Wheel, as per #1141 

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by tlambert03** [comment]
@DragaDoncila, what still needs to be done here? (haven't looked at the code)

**Comment by DragaDoncila** [comment]
@tlambert03 the functionality is all there, but we need to refactor a few things and add tests. I'll go through and comment where we need to refactor. I can take a look at this tomorrow, but if you'd like to push it through then of course, go for it.

File: napari/_qt/qt_viewer.py [hunk 1]
```diff
@@ -35,6 +36,13 @@
 from .._vispy import create_vispy_visual
 
 
+class OurSceneCanvas(SceneCanvas):
+    def _process_mouse_event(self, event):
+        if event.type == 'mouse_wheel' and len(event.modifiers) > 0:
+            return
+        super()._process_mouse_event(event)
+
```
Comment (DragaDoncila) [inline napari/_qt/qt_viewer.py:46]: We need to move this to a more appropriate spot - I'm not sure where would be best - qt_utils? It's own class? A new dedicated spot for monkey patching vispy?

File: napari/components/viewer_model.py [hunk 2]
```diff
@@ -94,8 +94,28 @@ def __init__(
         self.mouse_move_callbacks = []
         # Hold callbacks for when mouse is pressed, dragged, and released
         self.mouse_drag_callbacks = []
+        # Hold callbacks for when mouse wheel is scrolled
+
+        def _scroll(viewer, event):
+            from .._viewer_key_bindings import (
```
Comment (DragaDoncila) [inline napari/components/viewer_model.py]: This import breaks at top level as a circular import, so we should pull out increment_dims_left and increment_dims_right. I think it would work well as a dims method but I think @jni had some bones to pick with that?

Comment (jni) [reply napari/components/viewer_model.py]: I didn't have specific bones, I just wasn't sure. =) I think I was worried about the fact that dims will get refactored soon, but this will be affected either way. Thinking about it now, I agree, should be a Dims method with an axis kwarg and a None default, in which case last axis is used.

File: napari/_qt/qt_viewer.py (hunk 1 above)
Comment (jni) [reply napari/_qt/qt_viewer.py:46]: I think this is an ok spot for it since this is where we import SceneCanvas, but maybe we should rename it KeyFilterSceneCanvas or something descriptive like that? And the class docstring should describe the reason for its existence.

@DragaDoncila note that this is no longer monkey patching, it's subclassing, which is the Right and Good way to do this =), since we won't affect other running libraries using VisPy.

File: napari/components/viewer_model.py [hunk 3]
```diff
@@ -94,8 +94,28 @@ def __init__(
...
+
+            if 'Control' not in event.modifiers:
+                return
+            delta = int(event.delta[1])
+            if abs(delta) >= 1:
+                for i in range(abs(delta)):
+                    if delta < 0:
+                        increment_dims_left(self)
+                    else:
+                        increment_dims_right(self)
```
Comment (jni) [inline napari/components/viewer_model.py]: The last thing that needs to be done is that we need an accumulator for trackpad scrolling, which is fractional. So we would do something like:

```python
vertical_delta = event.delta[1]
int_delta = int(vertical_delta)  # vertical scrolling
fractional_delta = 
self._cumulative_delta += vertical_delta - int_delta
if self._cumulative_delta > 1:
    int_delta += 1
    self._cumulative_delta -= 1
```

... except there's two complicating factors:

- need to handle positive and negative deltas, which the above doesn't, and
- we need to reset cumulative delta to 0 after a certain time, or after any other event, since you don't really want the fractional scrolling from 3 minutes ago sitting around to modify your next scroll.

Probably I would say merge this without fractional deltas and then raise an issue for adding fractional deltas...

File: napari/components/viewer_model.py (hunk 2 above)
Comment (DragaDoncila) [reply napari/components/viewer_model.py]: So I tried pulling this out but I don't think it's that simple? If I put `increment_dims` on Dims, then we don't have access to `last_used`, because it belongs to `qt_dims`. If we put it in `qt_dims`, then we can't access it in `viewer_model` for the `_scroll` method. 

Also, regardless of which we put it in, we still need the `increment_dims_left/right` in the key bindings, in order to bind left and right arrows, so it would just call either `qt_dims`' or `dims`' method since it has access to both.

So now I'm not too sure where to put it. Left and right arrows works with it pretty much anywhere, but the scroll not so much.

Comment (sofroniewn) [reply napari/components/viewer_model.py]: > So now I'm not too sure where to put it. Left and right arrows works with it pretty much anywhere, but the scroll not so much.

For now we should probably put `last_used` on dims and remove it from QtDims, but we might want to think about another place longer term!

**Comment by DragaDoncila** [comment]
@ziyangczi @jni @sofroniewn I've added last_used as an attribute to dims in this last commit, and changed references from `qt_dims.last_used` to `dims.last_used`. The code to set and update `last_used` remains in `qt_dims`. All tests pass and scrolling (with and without modifier) works as expected from what I can tell. 

Still need to add a test, but if you'd like to take a look and make sure we're all on the same page that'd be great.

File: napari/components/viewer_model.py [hunk 4]
```diff
@@ -396,7 +396,7 @@ def _on_layers_change(self, event):
                 self.dims.set_range(i, r)
         self.events.layers_change()
 
-    def _scroll(self, event):
+    def _scroll(self, viewer, event):
```
Comment (jni) [inline napari/components/viewer_model.py:399]: Why is this change needed?

**Comment by jni** [comment]
@DragaDoncila very cool! Do you also want to tackle the fractional scrolling [issue](https://github.com/napari/napari/pull/1434#discussion_r456730417), or do you want to leave that for later?

File: napari/components/viewer_model.py (hunk 4 above)
Comment (DragaDoncila) [reply napari/components/viewer_model.py:399]: Without the viewer I got errors about the mouse callback expecting 2 arguments but receiving 3. When I looked at an older commit, we were passing the viewer so I added it back in

**Comment by DragaDoncila** [comment]
@jni totally happy to do the fractional scrolling, the only part I'm not sure about is resetting after a certain time/event. It needs to be done obviously but not sure what best approach would be. I'll leave it to you to decide if it should go in this PR or a follow up

File: examples/custom_mouse_functions.py [hunk 5]
```diff
@@ -16,6 +16,10 @@
     labeled = ndi.label(blobs)[0]
     labels_layer = viewer.add_labels(labeled, name='blob ID')
 
+    @viewer.mouse_press_callbacks.append
+    def get_event(viewer, event):
+        print(event)
+
```
Comment (jni) [inline examples/custom_mouse_functions.py:22]: This needs to go! 😂

Comment (jni) [reply examples/custom_mouse_functions.py:22]: No I take it back, I just saw it's in the examples 🤦 Clearly I need an afternoon ☕

File: napari/_qt/qt_viewer.py [hunk 6]
```diff
@@ -35,6 +36,17 @@
 from .._vispy import create_vispy_visual
 
 
+class KeyModifierFilterSceneCanvas(SceneCanvas):
+    """SceneCanvas which overrides default vispy behaviour for events with key modifiers.F
+
+    """
```
Comment (jni) [inline napari/_qt/qt_viewer.py]: PEP8/PEP257: one-liners should fit in one 79c line. 
    
```suggestion
    """SceneCanvas overriding VisPy when mouse wheel events have modifiers."""
```

**Review by jni — APPROVED** [review]

File: napari/components/viewer_model.py (hunk 4 above)
Comment (sofroniewn) [reply napari/components/viewer_model.py:399]: hmm this feels a little funny, i think i see why it was needed, but we might want to improve in a followup - this scroll method shouldn't need to live on the viewer model like this - the mixing of `self` and `viewer` seems like trouble and not something we've needed to do elsewhere.
